### PR TITLE
Add Codegen with Apollo

### LINKED ISSUE #6: Add Codegen with Apollo
Writing separate types by hand for every query is annoying - codegen is a feature that does it for you automatically

- Tutorial: https://www.apollographql.com/tutorials/lift-off-part1/09-codegen
- Example: https://github.com/apollographql/apollo-client-nextjs/blob/4930d28a971fa2397a6e225b9fb4b75744013216/examples/polls-demo/codegen.ts#L4

After all PRs are closed and we've separated the GitHub code to a separate repo, then let's integrate codegen with `@graphql-codegen/cli` as described in the

### PR DESCRIPTION
**Pull request by Eprince-hub**
Closes https://github.com/upleveled/graphql-example-fall-2023-atvie/issues/6

This PR adds codegen to the GraphQL example setup to generate TypeScript types based on our schema so that we don't have to write all the TypeScript types by hand

TODO
- [x] explore two setup possibilities
- [x] add sandboxes to both cases
- [x] implement most appropriate for our use case

Testing
- [x] CI Passes

The codegen implementation can be done in different ways, but the two methods described below are the ones done in. the documentation example

# The first example is the one that is given in the:
 [GraphQL Odyssey learning platform](https://www.apollographql.com/tutorials/lift-off-part1/09-codegen)
This method generates a lot of types in different files and types for the `gql` function that can be imported and used on the client.

## Pros
- Only two dependencies needed
- Generates complete type for every data point in GraphQl servers
- Provides `gql` function that can be imported and used (no need to import gql from `@apollo/client`)

## Cons
- Creates more types than necessary for our use case
- Creates multiple files for different types (Can be confusing)
- Generates complex types that may be difficult to understand/use

### Code Sandbox
[First Example](https://codesandbox.io/p/devbox/graphql-codegen-version-1-rygfks?file=%2Fpackage.json%3A1%2C1)

# The second example follows the implementation in the:
 [GraphQL example](https://github.com/apollographql/apollo-client-nextjs/blob/4930d28a971fa2397a6e225b9fb4b75744013216/examples/polls-demo/codegen.ts#L4)
This method generates all the needed types within a single file

## Pros
- Generates simple/readable types all in a single file
- Doesn't generate unnecessary types (Only the types from schema, mutations, and queries)
- Can be set to generate files and imports of types for all pages where GraphQL queries/mutations are run with (`preset: "near-operation-file"`) setting

## Cons
- Can need up to 6 dependencies (But I got the same functionalities with only three dependencies)
- Doesn't provide `gql` (Still need to import `gql` from `@apollo/client` for our queries and mutations)
- May lack the verbosity of the first implementation (Provides only the types needed)

### Code Sandbox
[Second Example](https://codesandbox.io/p/devbox/graphql-codegen-version-2-q652cy?file=%2Fcodegen.ts%3A13%2C27)

### CODE REVIEW

File: app/graphqlGeneratedTypes.ts
Comment (Eprince-hub) [inline app/graphqlGeneratedTypes.ts:1]: This TS file and code are generated when the codegen runs `pnpm generate`

File: codegen.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  schema: `${process.env.NEXT_PUBLIC_BASE_URL}/api/graphql`,
+  documents: ['app/**/*.tsx'],
+  generates: {
+    'app/graphqlGeneratedTypes.ts': {
+      plugins: ['typescript', 'typescript-operations'],
+    },
+  },
+};
+
+export default config;
```
Comment (Eprince-hub) [inline codegen.ts:17]: Codegen setup configuration that generates the code. Other options can be added as seen [here](https://github.com/apollographql/apollo-client-nextjs/blob/4930d28a971fa2397a6e225b9fb4b75744013216/examples/polls-demo/codegen.ts), but that requires installing more packages. This simple setup would be all we need for our use case

File: package.json [hunk 2]
```diff
@@ -17,7 +18,7 @@
     "@as-integrations/next": "^3.0.0",
     "@graphql-tools/schema": "^10.0.2",
     "@upleveled/ley": "^0.8.5",
-    "dotenv": "^16.4.1",
+    "dotenv-safe": "^9.0.0",
```
Comment (Eprince-hub) [inline package.json]: Replace dotenv with dotenv-safe as we start using it again after the issue it had was fixed

File: package.json [hunk 3]
```diff
@@ -30,6 +31,9 @@
     "zod": "^3.22.4"
   },
   "devDependencies": {
+    "@graphql-codegen/cli": "^5.0.2",
+    "@graphql-codegen/near-operation-file-preset": "^3.0.0",
+    "@graphql-codegen/typescript": "^4.0.4",
```
Comment (Eprince-hub) [inline package.json]: Codegen related dependencies

File: package.json [hunk 4]
```diff
@@ -8,7 +8,8 @@
     "dev": "next dev",
     "lint": "next lint",
     "migrate": "ley --require tsx/cjs",
-    "start": "next start"
+    "start": "next start",
+    "generate": "graphql-codegen"
```
Comment (Eprince-hub) [inline package.json]: Codegen related script

File: util/config.js [hunk 5]
```diff
@@ -1,23 +1,5 @@
...
-    dotenv.parse(readFileSync('./.env.example')),
-  ).filter((exampleKey) => !process.env[exampleKey]);
-
-  if (unconfiguredEnvVars.length > 0) {
-    throw new Error(
-      `.env.example environment ${
-        unconfiguredEnvVars.length > 1 ? 'variables' : 'variable'
-      } ${unconfiguredEnvVars.join(', ')} not configured in .env file`,
-    );
-  }
```
Comment (Eprince-hub) [inline util/config.js]: These codes no longer needed after switching to dotenv-safe

File: app/api/graphql/route.ts [hunk 6]
```diff
@@ -26,12 +33,6 @@ type GraphqlContext = {
   insecureSessionTokenCookie: RequestCookie | undefined;
 };
 
-type AnimalInput = {
-  firstName: string;
-  type: string;
-  accessory: string;
-};
-
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts:35]: Type no longer needed as the mutations are now using the types from the codegen
See addition of the types below

File: app/api/graphql/route.ts [hunk 7]
```diff
@@ -72,15 +73,15 @@ const resolvers = {
       return await getAnimalsInsecure();
     },
 
-    animal: async (parent: null, args: { id: string }) => {
+    animal: async (parent: null, args: QueryAnimalArgs) => {
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: Arguments now use the type from codegen

File: app/api/graphql/route.ts [hunk 8]
```diff
@@ -72,15 +73,15 @@ const resolvers = {
...
       return await getAnimalInsecure(Number(args.id));
     },
   },
 
   Mutation: {
     createAnimal: async (
       parent: null,
-      args: AnimalInput,
+      args: CreateAnimalMutationVariables,
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: Arguments now use the type from codegen

File: app/api/graphql/route.ts [hunk 9]
```diff
@@ -97,17 +98,16 @@ const resolvers = {
...
-        context.insecureSessionTokenCookie.value,
-        args.firstName,
-        args.type,
-        args.accessory,
-      );
+      return await createAnimal(context.insecureSessionTokenCookie.value, {
+        firstName: args.firstName,
+        type: args.type,
+        accessory: args.accessory || null,
+      });
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: The `createAnimal` database query function now accepts objects to correspond with the way query functions are written in the next.js example

File: app/api/graphql/route.ts [hunk 10]
```diff
@@ -97,17 +98,16 @@ const resolvers = {
...
     },
 
     deleteAnimal: async (
       parent: null,
-      args: { id: string },
+      args: DeleteAnimalMutationVariables,
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: Arguments now use the type from codegen

File: app/api/graphql/route.ts [hunk 11]
```diff
@@ -121,7 +121,7 @@ const resolvers = {
 
     updateAnimal: async (
       parent: null,
-      args: AnimalInput & { id: string },
+      args: UpdateAnimalMutationVariables,
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: Arguments now use the type from codegen

File: app/api/graphql/route.ts [hunk 12]
```diff
@@ -137,16 +137,15 @@ const resolvers = {
...
-        args.firstName,
-        args.type,
-        args.accessory,
-      );
+      return await updateAnimal(context.insecureSessionTokenCookie.value, {
+        id: Number(args.id),
+        firstName: args.firstName,
+        type: args.type,
+        accessory: args.accessory || null,
+      });
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts:144]: The `updateAnimal` database query function now accepts objects to correspond with the way query functions are written in the next.js example

File: app/api/graphql/route.ts [hunk 13]
```diff
@@ -137,16 +137,15 @@ const resolvers = {
...
     },
 
-    login: (parent: null, args: { username: string; password: string }) => {
+    login: (parent: null, args: LoginMutationVariables) => {
```
Comment (Eprince-hub) [inline app/api/graphql/route.ts]: Arguments now use the type from codegen

File: database/animals.ts [hunk 14]
```diff
@@ -25,12 +25,7 @@ export const getAnimalInsecure = cache(async (id: number) => {
 });
 
 export const createAnimal = cache(
-  async (
-    insecureSessionToken: string,
-    firstName: string,
-    type: string,
-    accessory: string,
-  ) => {
+  async (insecureSessionToken: string, newAnimal: Omit<Animal, 'id'>) => {
```
Comment (Eprince-hub) [inline database/animals.ts:29]: Database query functions are refactored to accept objects as it is in the next.js example and use the types from the migration files

File: database/animals.ts [hunk 15]
```diff
@@ -61,10 +56,7 @@ export const updateAnimal = cache(
   async (
     // FIXME: Rename insecureSessionToken to sessionToken everywhere
     insecureSessionToken: string,
-    id: number,
-    firstName: string,
-    type: string,
-    accessory: string,
+    updatedAnimal: Animal,
```
Comment (Eprince-hub) [inline database/animals.ts:60]: Database query functions are refactored to accept objects as it is in the next.js example and use the types from the migration files

File: app/graphqlGeneratedTypes.ts [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+
+export type UpdateAnimalMutation = { __typename?: 'Mutation', updateAnimal?: { __typename?: 'Animal', id: string, firstName: string, type: string, accessory?: string | null } | null };
+
+export type DeleteAnimalMutationVariables = Exact<{
+  id: Scalars['ID']['input'];
+}>;
+
+
+export type DeleteAnimalMutation = { __typename?: 'Mutation', deleteAnimal?: { __typename?: 'Animal', id: string } | null };
```
Comment (Eprince-hub) [inline app/graphqlGeneratedTypes.ts:113]: Codgen generated the types using the schema, queries and mutations as needed

File: codegen.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  schema: `${process.env.NEXT_PUBLIC_BASE_URL}/api/graphql`,
+  documents: ['app/**/*.tsx'],
+  generates: {
+    'app/graphqlGeneratedTypes.ts': {
+      plugins: ['typescript', 'typescript-operations'],
+    },
+  },
+};
+
+export default codegenConfig;
```
Comment (Eprince-hub) [inline codegen.ts]: Codegen setup configuration that generates the code. Other options can be added as seen [here](https://github.com/apollographql/apollo-client-nextjs/blob/4930d28a971fa2397a6e225b9fb4b75744013216/examples/polls-demo/codegen.ts), but that requires installing more packages. This simple setup would be all we need for our use case

File: database/connect.ts [hunk 18]
```diff
@@ -1,10 +1,10 @@
 import 'server-only';
+import { config } from 'dotenv-safe';
 import postgres from 'postgres';
-import { setEnvironmentVariables } from '../util/config.js';
 
 // This loads all environment variables from a .env file
 // for all code after this line
-setEnvironmentVariables();
+config();
```
Comment (Eprince-hub) [inline database/connect.ts:7]: The obsolete `setEnvironmentVariables` function removed, and import env configs directly from `dotenv-safe`

File: ley.config.js [hunk 19]
```diff
@@ -1,6 +1,6 @@
-import { setEnvironmentVariables } from './util/config.js';
+import { config } from 'dotenv-safe';
 
-setEnvironmentVariables();
+config();
```
Comment (Eprince-hub) [inline ley.config.js:3]: The obsolete `setEnvironmentVariables` function removed, and use env variables directly from `dotenv-safe`

Comment (karlhorky) [reply ley.config.js:3]: I think we need to keep the function (because of Vercel PostgreSQL env vars)

**Review by karlhorky — APPROVED** [review]

**Pull request merged**